### ISSUE TITLE
 Integrity constraint violation: 1062 Duplicate entry

### ISSUE THREAD

**Issue by esilvajr** [SammyK/LaravelFacebookSdk#127]
Using this package in 3.3 versin i get this error, Integrity constraint violation: 1062 Duplicate entry, qhen i call  App\User::createOrUpdateGraphNode($facebook_user).

He always try to insert new data and not execute a update in re-login.

Can anyone help?

**Comment by andreliem** [comment]
@esilvajr I have the exact same issue right now as well with 3.3.

Seems like there needs to be a way to tell createOrUpdateGraphNode what fields are "unique" and need to be used as lookups. For example, my problem is an integrity constraint with the email address, so I'm trying to see if I'm missing something with my configuration on telling LaravelFBSdk about what fields to use for uniqueness.

**Comment by andreliem** [comment]
So I think there's a big limitation with the library right now which should be patched. I hope I'm wrong but here's what I figured out.

```
  public static function createOrUpdateGraphNode($data)
    {
        // ...
        $attributes = [static::getGraphNodeKeyName() => $data['id']];

        $graph_node = static::firstOrNewGraphNode($attributes);
        // ...
    }
```

The $attributes here are hardcoded to rely on one identifier to determine if a User Model exists. In most cases, this would be the facebook_id field. So lets say you rely on an email address too, then this won't work. Every existing user who hasn't connected with Facebook but has a record with the same email that they have on Facebook will not be able to connect.

**Comment by SammyK** [comment]
Yes, the `createOrUpdateGraphNode()` method is quite basic and could certainly be upgraded to handle these type of scenarios better. I'll be diving into this lib some more when Laravel 5.3 comes out and will look into this more.

**Comment by andreliem** [comment]
@SammyK Thanks! I see that Laraavel 5.3 will have an updated API for createOrUpdate, seems like that will make this a lot easier.

**Comment by esilvajr** [comment]
@andreliem and @SammyK thank you for the answers! : )

**Comment by MeiRct** [comment]
@SammyK I found this open issue after experiencing the same behavior.
Laravel 5.3 is already out. Any chance you have time to please fix `createOrUpdateGraphNode()` method?

**Comment by andreliem** [comment]
@MeiRct In case you want a workaround, this is what I modified.

https://github.com/andreliem/LaravelFacebookSdk/blob/custom/src/LaravelFacebookSdk/SyncableGraphNodeTrait.php

Notice the extra parameter in this function
` public static function createOrUpdateGraphNode($data, $fieldId = 'id')`

Now you can specify other fields instead of relying just on the Facebook Id

**Comment by MeiRct** [comment]
A note from my insight, that might be helpful for others, and maybe matching also your case, @esilvajr.

I used @andreliem's workaround and still ran into "Duplicate entry" error.
Apparently the workaround was not for my case, because I'm using the 'id' column as the uniqueness lookup anyway.

After a little investigation, I found out I didn't use `bigIncrements` instead of `increments` for 'id' column in the migration file (should be: `$table->bigIncrements('id');`) , although @SammyK does mention in the repo docs to use unsigned big integer (https://github.com/SammyK/LaravelFacebookSdk#logging-the-user-into-laravel): 

> If you plan on using the Facebook user ID as the primary key, make sure you have a column called id that is an unsigned big integer and indexed.


Therefore, the database INSERT following the `save()` method was always inserting as the id the value of "4294967295", which is the maximum possible value for UNSIGNED INT in MySQL. 
But when the "SyncableGraphNodeTrait" trait called Larave's eloquent `firstOrNew()` method, it used the real retrieved facebook id for lookup, and always got new instance, because it couldn't find that id.

That's why it was consistently throwing "Duplicate entry" error, whenever the facebook id was bigger than 4294967295, because it attempted to insert a new record with id of 4294967295.
